src/proposal.mligo

```diff
@@ -277,84 +282,102 @@ let do_total_vote_meet_quorum_threshold (proposal, store: proposal * storage): b
-        let guardian = match dl_out.guardian with
-          | Some g -> g
-          | None -> store.guardian
-        in (dl_out.operations,
-              { store with extra = dl_out.extras
-              ; guardian = guardian
-              })
-      else (nil_op, store)
-    in
-    let cons = fun (l, e : operation list * operation) -> e :: l in
-    let ops = List.fold cons ops new_ops in
-    let store = remove_from_proposal_sort_by_level (start_level, proposal_key, store) in
-    (ops, store, counter)
-  else (ops, store, counter)
+    : (operation list * storage) =
+  let cond = do_total_vote_meet_quorum_threshold (proposal, store)
+          && proposal.upvotes > proposal.downvotes
+  in
+  let store = unstake_proposer_token
+        (config.rejected_proposal_slash_value, cond, proposal, config.period, config.fixed_proposal_fee_in_token, store) in
+  let (new_ops, store) =
+    if cond
+    then
+      let dl_out = config.decision_lambda { proposal = proposal; extras = store.extra } in
+      let guardian = match dl_out.guardian with
+        | Some g -> g
+        | None -> store.guardian
+      in (dl_out.operations,
+            { store with extra = dl_out.extras
+            ; guardian = guardian
+            })
+    else (nil_op, store)
+  in
+  let cons = fun (l, e : operation list * operation) -> e :: l in
+  let ops = List.fold cons ops new_ops in
+  (ops, store)
+
+
+(*
+ * Flush a proposal, starting from the first. Remove it when the flush is done, and move to the next proposal.
+ *)
+let rec flush_one (counter, config, store, ops : counter * config * storage * operation list): (operation list * storage * counter) =
+  // Ensure plist is not empty.
+  match store.ongoing_proposals_dlist with
+    | Some plist ->
+        // Ensure counter is not reached yet.
+        if (counter.current < counter.total) then
+          let proposal = check_if_proposal_exist (plist.first, store) in
+
+          // Ensure the proposal is not expired.
+          let _ = if is_proposal_age (proposal, config.proposal_expired_level)
+              then (failwith expired_proposal : unit)
+              else unit in
+          // Ensure the the proposal is flushable.
+          if (is_proposal_age (proposal, config.proposal_flush_level))
+            then
+              // Unstake tokens, and execute proposal's decision_lambda.
+              let (ops, store) = handle_proposal_is_over (config, proposal, store, ops) in
+
+              // Remove the first proposal from plist.
+              let store = { store with ongoing_proposals_dlist = plist_pop store.ongoing_proposals_dlist } in
+
+              // Continue to the next proposal
+              let counter = { counter with current = counter.current + 1n } in
+              flush_one (counter, config, store, ops)
+
+            else
+              (ops, store, counter)
+        else
+          (ops, store, counter)
+    | None ->
+        (ops, store, counter)
```

Comment:
If `plist_pop` returns the first element, then in a single call we can get the updated list and only need to care take care of the head key
```suggestion
let rec flush_one (counter, config, store, ops : counter * config * storage * operation list): (operation list * storage * counter) =
  let (plist_head, plist_new) = plist_pop store.ongoing_proposals_dlist in
  // Ensure plist is not empty.
  match plist_head with
    | Some first_key ->
        // Ensure counter is not reached yet.
        if (counter.current < counter.total) then
          let proposal = check_if_proposal_exist (first_key, store) in

          // Ensure the proposal is not expired.
          let _ = if is_proposal_age (proposal, config.proposal_expired_level)
              then (failwith expired_proposal : unit)
              else unit in
          // Ensure the the proposal is flushable.
          if (is_proposal_age (proposal, config.proposal_flush_level))
            then
              // Unstake tokens, and execute proposal's decision_lambda.
              let (ops, store) = handle_proposal_is_over (config, proposal, store, ops) in

              // Update the stored plist.
              let store = { store with ongoing_proposals_dlist = plist_new } in

              // Continue to the next proposal
              let counter = { counter with current = counter.current + 1n } in
              flush_one (counter, config, store, ops)
            else
              (ops, store, counter)
        else
          (ops, store, counter)
    | None ->
        (ops, store, counter)
```
which is a bit of an improvement.

BTW I think that rearranging the checks here might avoid us to doreturn `(ops, store, counter)` in 3 separate locations and improve this even more.

Reply:
Updated this, also re-arranged the check a bit, it returns `(ops, store, counter)` in 2 places now.